jib-cli/src/main/java/com/google/cloud/tools/jib/cli/buildfile/Layers.java

```diff
@@ -0,0 +1,179 @@
+                      (Predicate<Path>)
+                          path -> {
+                            // filter out against excludes
+                            for (PathMatcher matcher : excludes) {
+                              if (matcher.matches(path)) {
+                                return false;
+                              }
+                            }
+                            // if there are no includes directives, include everything
+                            if (includes.isEmpty()) {
+                              return true;
+                            }
+                            // TODO: for directories that fail to match the "include" directive on
+                            // TODO: files, populate the directories somehow or just never apply
+                            // TODO: includes to directories
+                            // if there are includes directives, only include those specified
+                            for (PathMatcher matcher : includes) {
+                              if (matcher.matches(path)) {
+                                return true;
+                              }
+                            }
+                            return false;
+                          })
+                  .map(
+                      path -> {
+                        Path relative = src.relativize(path);
+                        if (Files.isDirectory(path)) {
+                          return new FileEntry(
+                              path,
+                              dest.resolve(relative),
+                              filePropertiesStack.getDirectoryPermissions(),
+                              filePropertiesStack.getModificationTime(),
+                              filePropertiesStack.getOwnership());
+                        } else if (Files.isRegularFile(path)) {
+                          return new FileEntry(
+                              path,
+                              dest.resolve(relative),
+                              filePropertiesStack.getFilePermissions(),
+                              filePropertiesStack.getModificationTime(),
+                              filePropertiesStack.getOwnership());
+                        } else {
+                          throw new UnsupportedOperationException(
+                              "Cannot create FileLayers from non-file, non-directory: "
+                                  + path.toString());
+                        }
+                      })
+                  .forEach(layerBuiler::addEntry);
+            }
+          } else { // other
+            throw new UnsupportedOperationException(
+                "Cannot create FileLayers from non-file, non-directory: " + src.toString());
+          }
+          copySpec.getProperties().ifPresent(ignored -> filePropertiesStack.pop());
+        }
+        fileLayer.getProperties().ifPresent(ignored -> filePropertiesStack.pop());
+        layers.add(layerBuiler.build());
```

Comment:
Worth thinking about allowing to add an empty layer. But I think Jib will simply drop empty layers anyway, if I'm not mistaken.

Reply:
added todo, and bug in tracker bug

Reply:
Oh interesting, I didn't know empty layers were possible